### ISSUE TITLE
Build Menu UI ("Buildbar")

### ISSUE THREAD

**Issue by zverofaust** [SpringMCLegacy/SpringMCLegacy#17]
We need one of these to finalize the deconstruction of the legacy TA control panel.

We'll either need a multi-purpose buildbar or 3 (I think?) separate UI elements for the 3 different forms of "buying" units:

First, obviously, is ordering Mechs. I'd like to do away with the need for a "Submit Order" button. Or maybe we can just rename it to "Send Dropship" or something dramatic and cool. Deploy Dropship. Request Reinforcements. Whatever. 

Second is ordering Beacon Upgrades, Walls, Turrets and other Capture Point-related things (Logistics?).

Third is calling in aircraft sorties and orbital strikes and etc from the Orbital Relay (Supports?)

**Comment by gajop** [comment]
The dropship order menu and orbital relay should always be visible if available, and perhaps even combined into a single menu (although it's arguable if that's good). The orbital relay should only become available once built though.

**Comment by gajop** [comment]
I've been playing a bit with this and here's my input on this:
- I don't think it's important to separate mechs into light/medium/heavy/assault categories as it creates 4 extra buttons which clutter the interface, when there aren't actually that many different choices and it often makes sense to compare mechs between the same "class".
- Depending on the user knowledge a different display might make sense. Experienced players don't need detailed info but rather a fast interface.
- Build menu is the perfect place to integrate the dropship cooldown, which shouldn't be displayed all the time anyway!

With that, I'm showing the current WIP with 3 modes:
compact (default mode): http://i.imgur.com/HW3I1fi.png . shows important info about each mech 
small: http://i.imgur.com/oC9e9Ib.png . shows minimal information and allows the simultaneous display of multiple mechs (you can also see how I think the dropship cooldown should be displayed)
detail: http://i.imgur.com/HW3I1fi.png . shows detailed info about each mech and allows sorting (this was quickly hacked together for presentation)

All modes still have some major issues:
- No information about role is displayed (I'm not sure how to obtain this atm)
- The unit button colors and skin (image) are bad, as it's basically white on white and wastes too much space. I would prefer more crisp borders and maybe a different color scheme although this needs to be unified on the interface as a whole.
- Unit special abilities should also be displayed. I'm technically displaying the NARC button but none of the Steiner mechs seem to have it. I'm not sure if icons in bitmaps/icons are suitable for this.
- All UI elements should get additional and detailed tooltip information when mousing over, but there are some issues when using this chili font that cause spring to freeze (btw the font I think needs improvement, as it's hard to differentiate between C and O). Mousing over build icons should provide detailed info in all cases.

**Comment by FLOZi** [comment]
I disagree with not needing to categorise by weight, and that it is acceptable to display all mechs all the time.

Having said that, I would prefer a system of filters so you can quickly see e.g. all mechs with ECM, I can afford ~this~ one.

**Comment by gajop** [comment]
Well, while I can agree that showing all mechs might be an issue if that number is incredibly large, I'm not sure that grouping by weight class is the way to go. Roles/weapon types or a combination of multiple attributes are all valid ways of doing it. 
OTOH, a single way of grouping may be beneficial if key bindings are to be used.

Regarding fonts, I overrode the Chili theme to use this font everywhere. I'm not sure what your plans with fonts are yet.

**Comment by FLOZi** [comment]
Pretty much everything should use Handel Gothic, with 'headings' being in BT Oldstyle.

A role filter is useful too, but I think having a 'light mech' panel for the early game is too useful to discard. Any thoughts on what key bindings could be used?

**Comment by gajop** [comment]
I'm not saying using roles instead of weight classes is necessarily better. Unlike most RTSes, roles in MC are relatively fuzzy, and it's usually hard to tell what role a mech belongs to: particularly difficult being non-scout light mechs, sniper vs striker distinction, and similar. Even the classic LRM/ER-PPC Mad Cat variant is somewhere between a missile boat, sniper and striker.

In any case if groups are to be used, I think they should be fixed, but we don't have to decide right now on what that should be.
Key bindings will be really difficult seeing as there are quite a lot of mechs, but that can wait for now.

**Comment by gajop** [comment]
I haven't yet done anything regarding categories, but here's what I'm thinking with icons and tooltips:
Showing unit abilities: http://i.imgur.com/72pT6JT.png
Tooltips: http://i.imgur.com/m6ely59.png http://i.imgur.com/diGr8JG.png http://i.imgur.com/0zRE1l1.png

A couple of things: 
1. I think weapon range should be included in the name. I like MC1 style with just coloring weapons based on range (green, blue, red), thoughts?
2. There should be info on the weapon type (missile, direct fire projectile, energy/PPC), but weapondefs need to be populated with it first.
3. Unit role should be specified in the unitdefs and shown as an icon/colored text in the tooltip and if possible as icon in the build list.
4. Maybe HP and wepaon damage could be shown as well, but not sure it's relevant.

(Related issue: #41 )

**Comment by FLOZi** [comment]
Nice work, but re: tooltips -> https://github.com/SpringMCLegacy/SpringMCLegacy/wiki/Dev:-Preview-Card

1. I'm open to this but we need zvero's input
2. Weapon defs already have a "weaponclass" customparam that is/was used for perking, should be what you are after.
3. Unit description is just weightclass .. " " .. role
4. HP maybe, not sure weapon damage gives enough information in a single value to be relevant

**Comment by gajop** [comment]
After the games last night I really wish a better build menu existed.
I've seen the colors and shapes here: https://springrts.com/phpbb/viewtopic.php?f=70&t=31359&start=40#p553408 and I would prefer that but I think those can be added later on. I don't think it's a particularly good UI though as it doesn't show you anything about the unit on a glance.

I'd go with a GUI akin to http://i.imgur.com/72pT6JT.png . It's greatly simplified from what's being used now, and it shows key unit abilities. What it lacks is a clear indication of the unit's role, and more importantly, its range. Knowing what weapons the unit has (their name and even range) helps a bit, and I'd like to utilize the tooltip for that (http://i.imgur.com/0zRE1l1.png), but slightly simplify its display (I think one can simply say 4x ERPPC and color it by range. cooldown isn't that important really). I think part of it could be done via UnitCards, but chili tooltips don't support complex chili elements (yet). That said, I think it's imperative to add a role icon, so one can immediately tell the units role by looking at the list/grid (http://i.imgur.com/72pT6JT.png).

WRT separation by class - maybe, but I think it might be better to do that in the single window and separate each class with a line, akin to:
--- Light --- 
[] [] [] [] []
[] [] [] [] []
[] [] [] [] []
[] [] [] [] []
--- Medium ---
[] [] [] [] []
[] [] [] [] []
[] [] [] [] []
[] [] [] [] []
--- Heavy ---
[] [] [] [] []
[] [] [] [] []
...
The reasoning behind this is that it's faster to scroll than to click on tabs, and you often want to compare mechs of different class, trying to get the best thing you can afford.

**Comment by FLOZi** [comment]
I love hierarchy and tabs personally, +1 for the role icons, not a huge fan of tooltips but it will be good for newbies, so long as it is a seperate widget so i can turn it off.

If you build it, they will come, etc.

**Comment by FLOZi** [comment]
Reopening, still relevant as everyone but me hates the hacked up Spring-default build menu.

**Comment by zverofaust** [comment]
Alternative to a tooltip would be to simply show the highlighted Mech in the Unit Card, which shows weapons. I like the idea of little icons. Actually overall I like gajop's proposed GUI. There's even room to include things like total CBill cost of order and tonnage of order.

**Comment by FLOZi** [comment]
I actually think now with the role-based menu and proper buildpics with weapon type and special abilities indicated, the need for this is reduced. Also gajop is pretty busy with other Spring stuff and not likely to come back and work on this imo.
